### PR TITLE
Use MeshBase data to inform us when we should prepare

### LINKED ISSUE #15944: Prevent buggy simulation behavior due to lack of mesh preparedness
The meshgenerator system is currently missing calls to `prepare_for_use` and this can result in segfaults.

For example, this example with `GeneratedMeshGenerator` causes a segfault:
```
[Mesh]
  [./gmg]
    type = GeneratedMeshGenerator
    dim = 1
    nx = 1
    ny = 0
    nz = 0
    elem_type = EDGE2
  []

  [./extrude]
    type = FancyExtruderGenerator
    input = gmg
    heights = '1'
    num_layers = '1'
    direction = '0 1 0'
  []
[]

[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  [./bottom]
    type = DirichletBC
    variable = u
    boundary = '0'
    value = 0
  []
  [./top]
    type = DirichletBC
    variable = u
    boundary = '1'
    value = 1
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
[]

[Outputs]
  exodus = true
[]
```

Related issues: #15936 #15823

@friedmud and I spoke offline about this issue. He convinced me that we can't simply call `prepare_for_use` after each call to a `MeshGenerator::generate()` function. There needs to be some mechanism by which meshgenerators like `FancyExtruderGenerator` advertise that their output mesh is unprepared and other meshgenerators like `SideSetsBetweenSubdomainsGenerator` advertise that they need a prepared mesh.

### PR DESCRIPTION
**Pull request by lindsayad**
We also can signal to the `MeshBase` object if we've done some
operation that renders the mesh unprepared. If we do this meticulously
then we can trust checks of the `is_prepared` flag to tell us when we
need to `prepare_for_use` or not

Closes #15944 #15936 #15823

### CODE REVIEW

**Comment by lindsayad** [comment]
@smharper

**Comment by loganharbour** [comment]
Don't do it

**Comment by lindsayad** [comment]
The adaptivity test failure is unreal. This is a 2D mesh and the gold file has 2 sidesets and nodesets. WTF

**Comment by lindsayad** [comment]
This is ready for review. @smharper would you want to throw up a test related to some of your previous seg faults?

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -358,33 +354,28 @@ MooseMesh::freeBndElems()
...
 
-  if (dynamic_cast<DistributedMesh *>(&getMesh()) && !_is_nemesis)
-  {
-    // Call prepare_for_use() and don't mess with the renumbering
-    // setting
-    if (force || _needs_prepare_for_use)
-    {
-      CONSOLE_TIMED_PRINT("Preparing for use");
+  if (!dynamic_cast<DistributedMesh *>(&getMesh()) || _is_nemesis)
+    // For whatever reason we do not want to allow renumbering here nor ever in the future?
```
Comment (loganharbour) [inline framework/src/mesh/MooseMesh.C:364]: Any idea on what the logic is here?

I could understand it in some specific cases in testing... but that's about it. The more the contiguous the numbering the better, no?

**Comment by lindsayad** [comment]
For performance I think so yea, but there a *ton* of tests in BISON that fail if this is removed. And I guess I don’t care enough to figure out why...

**Comment by loganharbour** [comment]
> And I guess I don’t care enough to figure out why...

Fair enough. Definitely too big of a change to get mixed in with this.

In general, this will be good to me once we get a previously failing input in as a test.

**Comment by lindsayad** [comment]
Ok I've added @smharper 's example from #15944 that was previously seg-faulting.

**Review by loganharbour — DISMISSED** [review]

**Comment by lindsayad** [comment]
Going to merge this at the end of the day unless someone else reviews

File: framework/src/meshgenerators/SideSetsGeneratorBase.C [hunk 2]
```diff
@@ -64,6 +64,16 @@ SideSetsGeneratorBase::setup(MeshBase & mesh)
...
+
+  // Change this when we have more fine-grained control over advertising what we need need and how
+  // we satisfy those needs. For now we know we need to have neighbors...and we do have an explicit
+  // `find_neighbors` call...but we don't have a `neighbors_found` API and it seems off to do:
+  //
+  // if (!mesh.is_prepared())
+  //   mesh.find_neighbors()
+
+  if (!mesh.is_prepared())
+    mesh.prepare_for_use();
```
Comment (fdkong) [inline framework/src/meshgenerators/SideSetsGeneratorBase.C:76]: I was wondering why the mesh generator needs to call `mesh.prepare_for_use`?  Ideally, `MeshSetupCompleteAction` is the right guy to manage this.

**Review by fdkong — CHANGES_REQUESTED** [review]
Looks good for me.

File: framework/src/meshgenerators/SideSetsGeneratorBase.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/meshgenerators/SideSetsGeneratorBase.C:76]: This mesh generator requires that elements have properly had their neighbors set. See https://github.com/idaholab/moose/issues/15823

File: test/tests/meshgenerators/fancy_extruder_generator/need-neighbors.i [hunk 3]
```diff
@@ -0,0 +1,56 @@
+[Mesh]
+  [./gmg]
+    type = GeneratedMeshGenerator
+    dim = 1
+    nx = 1
+    ny = 0
+    nz = 0
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply test/tests/meshgenerators/fancy_extruder_generator/need-neighbors.i:7]: Haha I just copied and pasted this from @smharper's issue. I can remove these lines

File: framework/src/meshgenerators/SideSetsGeneratorBase.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/meshgenerators/SideSetsGeneratorBase.C:77]: I added to the code comment, about how we need neighbors properly set, that the interested party should look at #15823

**Comment by lindsayad** [comment]
@fdkong does this look ok?

**Review by fdkong — APPROVED** [review]

**Pull request merged**